syntax = "proto3";
package proto;
option go_package = "./proto";

message Empty {}

message RegisterRequest {
  string store_server_address = 1;
}

enum Category {
  MISC = 0;
  SCANNER = 1;
  PASSWORD = 2;
  SHELL = 3;
  EXPLOIT = 4;
  WEB = 5;
  SOCIAL_ENGINEERING = 6;
  FORENSIC = 7;
  REPORTING = 8;
}

message RegisterResponse {
  message Route {
    string method = 1;
    string path = 2;
    bool use_sse = 3;
  }
  string id = 1;
  string name = 2;
  string version = 3;
  repeated Route routes = 4;
  Category category = 5;
  message Metadata {
    string key = 1;
    string value = 2;
  }
  repeated Metadata metadata = 6;
}

message Header {
  message Value {
    repeated string values = 1;
  }
  map<string, Value> header = 1;
}

message Request {
  string method = 1;
  string url = 2;
  Header header = 3;
  bytes body = 4;
}

message Response {
  int32 status = 1;
  Header header = 2;
  string body = 3;
}

service Module {
  rpc Register(RegisterRequest) returns (RegisterResponse);
  rpc Handle(Request) returns (Response);
  rpc HandleSSE(Request) returns (stream Response);
}